### ISSUE TITLE
Support for proxy hosts black listing in ProxyOptions

### ISSUE THREAD

**Issue by haithkris** [eclipse-vertx/vert.x#2600]
Hi, we have this issue when we want to ignore the proxy and when it comes to call certain components of our internal network. Since vertx ignores JVM proxy arguments like `http.nonProxyHosts` there is no real workaround.

The proxy configuration is set at the `WebClient` level. ProxyOptions object should have an equivalent to `http.nonProxyHosts`.

Vertx web version: 3.5.1
Vertx core version: 3.5.1

**Comment by vietj** [comment]
is that something you can contribute ?

**Comment by haithkris** [comment]
I would if had proper time and proper knowledge on how vertx works in the internal. But I was hoping you guys can pull it through.

**Comment by vietj** [comment]
ok, i'm keeping this as feature request

**Comment by marcottedan** [comment]
Actually I believe it's more "whitelisting" than "blacklisting" since the noProxy feature of the JVM (and docker network has the same feature) is that everything goes to the proxy except the `noProxy` hosts.

**Comment by KowalczykBartek** [comment]
Hi, I crafted something today, 
of course there is a lot of to do in this PR still (only HttpClientImpl covered), but I wanted to ask what do you think about API, 
cc  @haithkris
```java
new ProxyOptions()
  .setType(ProxyType.HTTP)
  .exludeHost("localhost") //
  .exludeHost("localhost2") //
  .setHost("google.com")
  .setPort(proxy.getPort());
```

**Comment by haithkris** [comment]
looks good for me!

**Comment by vietj** [comment]
the setter should be `setExcludedHosts(List<String>)` and there should be a method `addExcludedHost(String)` that adds element to the list.

Look at how other options handle this.

**Comment by KowalczykBartek** [comment]
What with NetClient ? should ProxyOptions include option to exclude SocketAddress ?
```java
NetClient connect(SocketAddress remoteAddress, Handler<AsyncResult<NetSocket>> connectHandler);
```

**Comment by vietj** [comment]
that's a concern I had indeed, but ProxyOptions don't support `@VertxGen` so I think the best would be to have the excluded host to begin with `/` instead and we should parse them accordingly

**Comment by KowalczykBartek** [comment]
> so I think the best would be to have the excluded host to begin with / instead and we should parse them accordingly when we build the proxy

can you explain it a little bit ?

**Comment by vietj** [comment]
well actually I don't think we need to do that because one can still create 2 net clients if he wants to use domain sockets.

so just a plain list of excluded hosts is fine for me, we don't need to address the port.

**Comment by KowalczykBartek** [comment]
hmm, so, I guess that I can ask you for review,
If someone wants to test it "in action", I played as follow: 

```java
CountDownLatch a = new CountDownLatch(1);

Vertx v = Vertx.vertx();

ProxyOptions proxyOptions = new ProxyOptions();
proxyOptions.setHost("targethost");
proxyOptions.setPort(8080);
proxyOptions.addExcludedHost("exludedhost1");
proxyOptions.addExcludedHost("exludedhost2");

HttpClient client = v.createHttpClient(new HttpClientOptions()
        .setProxyOptions(proxyOptions));

client.getNow(8080,"exludedhost1","/", request -> {
    System.out.println(request);
    a.countDown();
});

a.await();
```
and /etc/hosts
```
127.0.0.1 targethost
127.0.0.1 exludedhost1
127.0.0.1 exludedhost2
```

P.S As I played with this proxy feature, I noticed that it doesn't work with HTTPS traffic, ProxyChannelProvider do not understand https, generates HTTP request, then this proxied http request is handled by handler(SslHandler) that expects to have SSL record. 
this is request that fails
```
ProxyOptions proxyOptions = new ProxyOptions();
proxyOptions.setHost("targethost");
proxyOptions.setPort(8080);
proxyOptions.addExcludedHost("exludedhost1");
proxyOptions.addExcludedHost("exludedhost2");

HttpClient client = v.createHttpClient(new HttpClientOptions()
        .setSsl(true)
        .setTrustAll(true)
        .setVerifyHost(false)
        .setProxyOptions(proxyOptions));
```
did I do something wrong ? maybe I can fix this in this branch as well ?

**Comment by vietj** [comment]
@KowalczykBartek it's not clear what you mean with https. The HttpClient supports HTTPS proxy using the connect method, and the traffic will be encrypted after the proxy issues an CONNECT request to the server. This is explained here https://vertx.io/docs/vertx-core/java/#_using_a_proxy_for_http_https_connections

**Comment by vietj** [comment]
I think what you are missing is the `ChannelConnect#doConnect` method that here will make the decision to use or not a proxy:

```
    if (options.getProxyOptions() == null || !ssl && options.getProxyOptions().getType()== ProxyType.HTTP ) {
      channelProvider = ChannelProvider.INSTANCE;
    } else {
      channelProvider = ProxyChannelProvider.INSTANCE;
    }
```

**Comment by vietj** [comment]
the boolean usesProxy is only used for non https proxies that will rewrite the URL for the proxy as a full uri

**Comment by KowalczykBartek** [comment]
That is question to @haithkris , i just wanted to make some code :)

**Comment by vietj** [comment]
oh right, if you are looking for contributions, we can guide you and assign you tasks

**Comment by vietj** [comment]
@marcottedan the JVM has a global handling (static), in vertx you can create (lightweight) many different clients with different configurations, so it's fine to create one HttpClient with proxy settings and one HttpClient without proxies and let the application chose the right client

**Comment by KowalczykBartek** [comment]
@vietj can you please check email (sorry that I am pinging here XD ) ?

**Comment by KowalczykBartek** [comment]
@vietj do you mean some stuff not present/listed in Issues vert.x tab ?

**Comment by vietj** [comment]
@KowalczykBartek I got your mail and will reply to you soon, sorry for the delay

**Comment by laurentvaills** [comment]
Hi guys. 
Any news on this feature ? It seems unfortunately  inactive since Sept 2018. 
Cheers.

**Comment by vietj** [comment]
we should review it and merge it I think

**Comment by laurentvaills** [comment]
Thanks @vietj .

**Comment by cescoffier** [comment]
@vietj any update on this one?

**Comment by vietj** [comment]
it is not yet scheduled to be implemented by the team

**Comment by gurusreekanth** [comment]
Hi Julien,

Is there any plans to port this fix to 3.9.x releases?

Thank you.

**Comment by vietj** [comment]
there are no plans, we only support 3.9.x for bug fixes, sorry
